src/websocket.js

```diff
@@ -0,0 +1,259 @@
+      this.latestRTT = dt;
+      this.latestPingDate = null;
+    }
+    if (this.timeoutTimer) {
+      clearTimeout(this.timeoutTimer);
+      this.timeoutTimer = null;
+    }
+  }
+
+  /**
+   * Handle message receiving from websocket
+   *
+   * @param {Object} evt Event that has data (evt.data) sent in the websocket
+   */
+  onMessage(evt) {
+    const message = JSON.parse(evt.data)
+    const _type = message.type.split(':')[0]
+    if (_type === 'pong') {
+      this.onPong();
+    }
+    this.emit(_type, message)
+  }
+
+  /**
+   * Method called when websocket connection is opened
+   */
+  onOpen() {
+    this.connected = true;
+    this.started = true;
+    this.setIsOnline(true);
+    this.heartbeat = setInterval(() => {
+      this.sendPing();
+    }, this.heartbeatInterval);
+  }
+
+  /**
+   * Method called when websocket connection is closed
+   */
+  onClose() {
+    this.started = false;
+    this.connected = false;
+    this.setIsOnline(false);
+    if (this.ws) {
+      this.ws.close();
+      this.ws = null;
+    }
+    setTimeout(() => {
+      this.setup()
+    }, this.retryConnectionInterval);
+    clearInterval(this.heartbeat);
+  }
+
+  /**
+   * Method called when an error happend on websocket
+   *
+   * @param {Object} evt Event that contains the error
+   */
+  onError(evt) {
+    this.onClose();
+    // console.log('ws error', window.navigator.onLine, evt);
+  }
+
+  /**
+   * Method called to send a message to the server
+   *
+   * @param {string} msg Message to be sent to the server (usually JSON stringified)
+   */
+  sendMessage(msg) {
+    if (!this.started) {
+      this.setIsOnline(false);
+      return;
+    }
+
+    if (this.ws.readyState === WS_READYSTATE_READY) {
+      this.ws.send(msg);
+    } else {
+      // If it is still connecting, we wait a little and try again
+      setTimeout(() => {
+        this.sendMessage(msg);
+      }, 1000);
+    }
+  }
+
+  /**
+   * Ping method to check if server is still alive
+   *
+   */
+  sendPing() {
+    if (this.latestPingDate) {
+      // Skipping sendPing. Still waiting for pong...
+      return;
+    }
+    const msg = JSON.stringify({'type': 'ping'})
+    this.latestPingDate = new Date();
+    this.timeoutTimer = setTimeout(() => this.onConnectionDown(), this.connectionTimeout);
+    this.sendMessage(msg)
+  }
+
+  onConnectionDown() {
+    this.onClose();
+  };
+
+  /**
+   * Method called to end a websocket connection
+   *
+   */
+  endConnection() {
+    this.setIsOnline(undefined);
+    this.started = false;
+    this.connected = undefined;
+    if (this.ws) {
+      this.ws.onclose = () => {};
+      this.ws.close();
+      this.ws = null;
+    }
+    clearInterval(this.heartbeat);
+  }
+
+  /**
+   * Set in redux if websocket is online
+   *
+   * @param {*} value Can be true|false|undefined
+   */
+  setIsOnline(value) {
+    // Save in redux
+    // Need also to keep the value in 'this' because I was accessing redux store
```

Comment:
Is it an outdated comment? I think this library shouldn't depend on redux (even though we can have a `redux.js` to make redux integration easier).

Reply:
Yes, it's an old code. We used to have this code in the wallet (where we use redux). Removed it